OSDOCS-11158: adds image mode to MicroShift install (#80599)
openshift/openshift-docs: modules/microshift-install-rhel-image-mode-build-image.adoc

// Module included in the following assemblies:
//
// microshift_install_bootc/microshift-install-rhel-image-mode.adoc

:_mod-docs-content-type: PROCEDURE
[id="microshift-rhel-image-mode-build-image_{context}"]
= Building the bootc image

Build your {op-system-base-full} that contains {microshift-short} as a bootable container image by using a Containerfile.

[NOTE]
====

> Image mode for {op-system} is technology preview. Using a bootc image in production environments is not supported.

Reviewer: ```suggestion
Image mode for {op-system} is Technology Preview. Using a bootc image in production environments is not supported.
```

====

.Prerequisites
* A {op-system-base-full} 9.4 host with an active Red Hat subscription for building {microshift-short} bootc images and running containers.
* You are logged into the {op-system} 9.4 host using the user credentials that have `sudo` permissions.
* The `rhocp` and `fast-datapath` repositories are accessible in the host subscription, but do not necessarily need to be enabled on the host.
* You have a remote registry such as link:https://quay.io[Red Hat quay] for storing and accessing bootc images.

.Procedure

> . Create a Containerfile that includes the following instructions:

Author: I don't think so, see https://docs.podman.io/en/v3.0/markdown/podman-build.1.html and https://docs.openshift.com/container-platform/4.16/machine_configuration/mco-coreos-layering.html.

// Module included in the following assemblies:
//
// microshift_install_bootc/microshift-install-rhel-image-mode.adoc

:_mod-docs-content-type: PROCEDURE
[id="microshift-rhel-image-mode-build-image_{context}"]
= Building the bootc image

> Build your {op-system-base-full} that contains {microshift-short} as a bootable container image by using a a Containerfile.

Reviewer 2: ```suggestion
Build your {op-system-base-full} that contains {microshift-short} as a bootc image by using a a Containerfile.
```

Reviewer 2: We have been advised to use "bootc image" instead of "bootable containers".

Reviewer 3: ```suggestion
Build your {op-system-base-full} that contains {microshift-short} as a bootc image by using a Containerfile.
```

> Build your Red Hat Enterprise Linux as a bootable container image by using a `ContainerFile` and include {microshift-short}.

Reviewer 3: Can we rephrase this? Let's make it more explicit that the image we build includes MicroShift.

[NOTE]
====

> Image mode for {op-system-base-full} is technology preview. Using in production environments is not supported.

Reviewer 3: ```suggestion
Image mode for {op-system-base-full} is technology preview. Using it in production environments is not supported.
```

====

.Prerequisites
* A {op-system-base-full} 9.4 host with an active Red Hat subscription for building {microshift-short} bootc images and running containers.

> * You are logged into the {op-system} 9.4 host using the user credentials that have SUDO permissions configured.

Author: OpenShift uses direct root sometimes, and `sudo` other times. See https://docs.openshift.com/container-platform/4.16/installing/installing_ibm_cloud/install-ibm-cloud-installation-workflow.html. MicroShift itself doesn't have roles, so we prefer `sudo` in most cases.

* The `rhocp` and `fast-datapath` repositories are accessible in the host subscription, but do not necessarily need to be enabled on the host.

> * You have a remote registry such as link:https:/quay.io[Red Hat quay] for storing and accessing bootc images.

Reviewer 3: ```suggestion
* You have a remote registry such as link:https://quay.io[Red Hat quay] for storing and accessing bootc images.
```

.Procedure

> . Create a Containerfile and include the following instructions:

Reviewer 3: include following instructions? 
```suggestion
. Create a Containerfile including the following commands:
```

Author: Yes, per the RHEL doc team review. It is consistent with how they talk about this, so if the user goes there for additional information, it will be the same. WDYT?

Reviewer 3: It sounds strange to me, but OK, if they insist

Author: It looks like it's in the Podman docs as well, so I that seems to be the RH way. Okie dokie!

[source,text]
----
FROM registry.redhat.io/rhel9/rhel-bootc:9.4

ARG USHIFT_VER=4.16
RUN dnf config-manager \
        --set-enabled rhocp-${USHIFT_VER}-for-rhel-9-$(uname -m)-rpms \
        --set-enabled fast-datapath-for-rhel-9-$(uname -m)-rpms
RUN dnf install -y firewalld microshift && \
    systemctl enable microshift && \
    dnf clean all

# Create a default 'redhat' user with the specified password.
# Add it to the 'wheel' group to allow for running sudo commands.
ARG USER_PASSWD
RUN if [ -z "${USER_PASSWD}" ] ; then \
        echo USER_PASSWD is a mandatory build argument && exit 1 ; \
    fi
RUN useradd -m -d /var/home/redhat -G wheel redhat && \
    echo "redhat:${USER_PASSWD}" | chpasswd

# Mandatory firewall configuration
RUN firewall-offline-cmd --zone=public --add-port=22/tcp && \
    firewall-offline-cmd --zone=trusted --add-source=10.42.0.0/16 && \
    firewall-offline-cmd --zone=trusted --add-source=169.254.169.1
# Application-specific firewall configuration
RUN firewall-offline-cmd --zone=public --add-port=80/tcp && \
    firewall-offline-cmd --zone=public --add-port=443/tcp && \
    firewall-offline-cmd --zone=public --add-port=30000-32767/tcp && \

>     firewall-offline-cmd --zone=public --add-port=30000-32767/udp

Reviewer 3: Let's remove this block and leave it to the users to configure firewall according to their requirements.

Author: Left the mandatory, removed application-specific.

# Create a systemd unit to recursively make the root filesystem subtree
# shared as required by OVN images
RUN cat > /etc/systemd/system/microshift-make-rshared.service <<'EOF'
[Unit]
Description=Make root filesystem shared
Before=microshift.service
ConditionVirtualization=container
[Service]
Type=oneshot
ExecStart=/usr/bin/mount --make-rshared /
[Install]
WantedBy=multi-user.target
EOF
RUN systemctl enable microshift-make-rshared.service
----
+
[IMPORTANT]
====

> Podman uses the host subscription information and repositories inside the container when building the container image. If the `rhocp` and `fast-datapath` repositories are not available, the build fails.

Reviewer 3: ```suggestion
Podman uses the host subscription information and repositories inside the container when building the container image. If the `rhocp` and `fast-datapath` repositories are not available on the host, the build fails.
```

// Module included in the following assemblies:
//
// microshift_install_bootc/microshift-install-rhel-image-mode.adoc

:_mod-docs-content-type: PROCEDURE
[id="microshift-rhel-image-mode-build-image_{context}"]
= Building the bootc image

Build your Red Hat Enterprise Linux as a bootable container image by using a `ContainerFile` and include {microshift-short}.

[NOTE]
====
Image mode for {op-system-base-full} is technology preview. Using in production environments is not supported.
====

.Prerequisites
* A {op-system-base-full} 9.4 host with an active Red Hat subscription for building {microshift-short} bootc images and running containers.
* You are logged into the {op-system} 9.4 host using the user credentials that have SUDO permissions configured.

> * The `rhocp` and `fast-datapath` repositories are enabled so that {microshift-short} and its dependencies can be installed. These repositories must be accessible in the host subscription, but do not necessarily need to be enabled on the host.

Reviewer 3: ```suggestion
* The `rhocp` and `fast-datapath` repositories are accessible in the host subscription, but do not necessarily need to be enabled on the host.
```

* You have a remote registry such as link:https:/quay.io[Red Hat quay] for storing and accessing bootc images.

> //Q: do we need this one since we aren't documenting running the VM?

Reviewer 3: We need this because we're documenting pushing an image to a registry

> * A physical hypervisor host with the `libvirt` virtualization platform for starting virtual machines that run {op-system} operating system containing {microshift-short} binaries. See link:https://libvirt.org/[libvirt Virtualization API] for more information.

Reviewer 3: This one is not necessary as we're not start VMs

.Procedure

//Do we want to give a command for this?
//Does it matter where the file is or what the name is?

> . Create a `ContainerFile` and include the following contents:

Reviewer 3: File name matters because it's used subsequently in commands.
```suggestion
. Create a `Containerfile` file and include the following contents:
```

+

> .Example ContainerFile for {op-system} image mode

Reviewer 3: ```suggestion
.Example Containerfile for {op-system} image mode
```

Author: Ah, I see. OpenShift docs do not backtick this, nor do the RHEL docs, so we should be good to follow those styles. See https://docs.openshift.com/container-platform/4.16/machine_configuration/mco-coreos-layering.html#coreos-layering-examples_mco-coreos-layering

        echo USER_PASSWD is a mandatory build argument && exit 1 ; \
    fi
RUN useradd -m -d /var/home/redhat -G wheel redhat && \
    echo "redhat:${USER_PASSWD}" | chpasswd

# Mandatory firewall configuration
RUN firewall-offline-cmd --zone=public --add-port=22/tcp && \
    firewall-offline-cmd --zone=trusted --add-source=10.42.0.0/16 && \
    firewall-offline-cmd --zone=trusted --add-source=169.254.169.1
# Application-specific firewall configuration
RUN firewall-offline-cmd --zone=public --add-port=80/tcp && \
    firewall-offline-cmd --zone=public --add-port=443/tcp && \
    firewall-offline-cmd --zone=public --add-port=30000-32767/tcp && \
    firewall-offline-cmd --zone=public --add-port=30000-32767/udp

# Create a systemd unit to recursively make the root filesystem subtree
# shared as required by OVN images
RUN cat > /etc/systemd/system/microshift-make-rshared.service <<'EOF'
[Unit]
Description=Make root filesystem shared
Before=microshift.service
ConditionVirtualization=container
[Service]
Type=oneshot
ExecStart=/usr/bin/mount --make-rshared /
[Install]
WantedBy=multi-user.target
EOF
RUN systemctl enable microshift-make-rshared.service
----
+
[IMPORTANT]
====

> Podman uses the host subscription information and repositories inside the container when building the container image. If the `rhocp` and `fast-datapath` repositories are not enabled, the build fails.

Reviewer 3: ```suggestion
Podman uses the host subscription information and repositories inside the container when building the container image. If the `rhocp` and `fast-datapath` repositories are not available, the build fails.
```
